### ISSUE TITLE
hOCR renderer could perhaps write scan_res to the ocr_page metadata

### ISSUE THREAD

**Issue by MerlijnWajer** [tesseract-ocr/tesseract#3326]
### Environment

* **Tesseract Version**: 5.0.0-alpha-20201231

### Current Behavior:

I've found that Tesseract will scale the `x_fsize` if the DPI is passed at the command line (via `--dpi`) or if the DPI if embedded in the metadata of the image itself (exif tags, etc). However, it will not do this when it attempts to estimate the DPI. All of this is fine and makes sense to me, but for a consumer of the hOCR file, the actual DPI that was used to scale the x_fsize is nowhere to be found. That is: Tesseract writes the `x_fsize` in printer points, but doesn't specify what it thought the DPI of the image was.

### Expected Behavior:

Tesseract provides an entry in the hOCR files that specifies the identified DPI that was used to scale various values in the hOCR file.

### Suggested Fix:

The hOCR spec has a `scan_res` field for the `ocr_page` entry: http://kba.cloud/hocr-spec/1.2/#scan_res

Using this `scan_res` field, consumers of the hOCR output file can construct the size of the text in pixels.

I could conjure up a patch to include this property in the hOCR output of every `ocr_page` element, but *only* if it was detected in the image metadata, or passed via the command line. In other cases, Tesseract will guess what the DPI is, but does not actually adjust the `x_fsize` attribute.

**Comment by MerlijnWajer** [comment]
I currently went for this. Not sure what to do with the x_res.

```
From 9f82e08f84e74737d2b8a13f6b8fbc2105aef11b Mon Sep 17 00:00:00 2001
From: Merlijn Wajer <merlijn@wizzup.org>
Date: Mon, 15 Mar 2021 19:31:25 +0100
Subject: [PATCH] hocrrenderer: write scan_res

---
 src/api/hocrrenderer.cpp | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/src/api/hocrrenderer.cpp b/src/api/hocrrenderer.cpp
index f4504da9..738de926 100644
--- a/src/api/hocrrenderer.cpp
+++ b/src/api/hocrrenderer.cpp
@@ -175,9 +175,11 @@ char* TessBaseAPI::GetHOCRText(ETEXT_DESC* monitor, int page_number) {
   } else {
     hocr_str << "unknown";
   }
-  hocr_str << "\"; bbox " << rect_left_ << " " << rect_top_ << " "
-           << rect_width_ << " " << rect_height_ << "; ppageno " << page_number
-           << "'>\n";
+
+  hocr_str << "\"; bbox " << rect_left_ << " " << rect_top_ << " " << rect_width_ << " "
+           << rect_height_ << "; ppageno " << page_number
+           << "; scan_res " << thresholder_->GetSourceYResolution() << " "
+           << thresholder_->GetSourceYResolution() << "'>\n";
 
   std::unique_ptr<ResultIterator> res_it(GetIterator());
   while (!res_it->Empty(RIL_BLOCK)) {
-- 
```

**Comment by MerlijnWajer** [comment]
Any comments on this? Shall I turn this into a pull request?

**Comment by MerlijnWajer** [comment]
@stweil or @amitdo  - do you have any objection to this change? I think the hOCR output is degraded quite a bit by the fact that it just blows up the font sizes with no indication of the DPI that was being used. I've been adding this patch to every Tesseract build I do because it's quite important for PDF generation down the line.

**Comment by MerlijnWajer** [comment]
I would argue it might be a bug rather than a feature request, too. :-)

**Comment by zdenop** [comment]
Could you send a PR against the latest code?

**Comment by amitdo** [comment]
Note that there are upto 3 DPIs used in Tesseract. DPI from the image, user provided DPI, Tesseract calculated DPI.

**Comment by MerlijnWajer** [comment]
@zdenop - will do this weekend

@amitdo - right. My understanding is that the value that is written by this PR is the value that Tesseract reads from the image (metadata), or what is provided on the command line (user dpi), because that's what is used to scale the font size. The estimated DPI is not used for that purpose. I can test if the user-provided DPI takes precedence.

**Comment by stweil** [comment]
What is done if there is neither a DPI in the metadata of the image nor provided on the command line?

**Comment by amitdo** [comment]
IIRC, if the DPI value is less than `70` or not provided, Tesseract will use `70` as the value for the font scaling calculation.

**Comment by MerlijnWajer** [comment]
In that case the default dpi is written, which I think is 70.

Example of image with dpi (400) written to the JPEG2000 metadata (and picked up by Leptonica I think):

```
$ TESSDATA_PREFIX=/usr/share/tessdata LD_LIBRARY_PATH=`pwd` LD_LIBRARY_PATH=$LD_LIBARY_PATH:`pwd`/.libs ~/archive/tesseract-src/tesseract/.libs/tesseract  -c hocr_font_info=1 /tmp/sim_multinational-monitor_september-october-2006_27_5_0042.jp2 - hocr | egrep '(consensus|scan_res)'
Detected 84 diacritics
  <div class='ocr_page' id='page_1' title='image "/tmp/sim_multinational-monitor_september-october-2006_27_5_0042.jp2"; bbox 0 0 2952 4000; ppageno 0; scan_res 400 400'>
      <span class='ocrx_word' id='word_1_19' title='bbox 445 471 634 492; x_wconf 95; x_fsize 8'>consensus</span>
```

With the same image, but converted to a `bmp` to strip the metadata:

```
$ TESSDATA_PREFIX=/usr/share/tessdata LD_LIBRARY_PATH=`pwd` LD_LIBRARY_PATH=$LD_LIBARY_PATH:`pwd`/.libs ~/archive/tesseract-src/tesseract/.libs/tesseract  -c hocr_font_info=1 /tmp/sim_multinational-monitor_september-october-2006_27_5_0042.bmp - hocr | egrep '(consensus|scan_res)'
Estimating resolution as 279
  <div class='ocr_page' id='page_1' title='image "/tmp/sim_multinational-monitor_september-october-2006_27_5_0042.bmp"; bbox 0 0 2952 4000; ppageno 0; scan_res 70 70'>
      <span class='ocrx_word' id='word_1_20' title='bbox 445 471 634 492; x_wconf 95; x_fsize 45'>consensus</span>
```

(Also note the difference in `x_fsize`)

**Comment by wollmers** [comment]
@MerlijnWajer 

Looks like Tesseracts measures `x_fsize` in pixels, if the `dpi` is estimated. In the second example the height of the `bbox` is 492 - 471 = 21 pixels for "consensus" (small characters only). Thus `x_fsize 45` is reasonable as the `x-height` in most Roman regular fonts is ~50% of font-size. 21/45 ~ 47%. 45 pixels at 70 dpi would be 43.43 points (European metric points; 1 point = 0.376 mm). Otherwise 45 pixel at dpi 400 gives 7.60 points or rounded 8 points.

The estimated `x_fsize` is not very reliable like in this example with base-, ascender-, x- and descender-lines drawn:

![Bildschirmfoto 2021-09-19 um 14 04 08](https://user-images.githubusercontent.com/1275557/133926818-d74066d7-857a-45d0-8da2-97602db3989b.png)

For the first line `x_size 52` -> `x_fsize 12` at 300 dpi is correct.

The second line has a height of 44, thus `x_fsize 12` is wrong and should be 10 (9.91).

The third line has a height of 50 giving 11.26 points, which maybe are the estimated `x_fsize 12` as the digit 8 does not use the full descender.

IMHO the issue is to have `x_fsize` in points in one case and in pixels in the other case, without specifying the units.

**Comment by MerlijnWajer** [comment]
@wollmers - my understanding is that the estimated DPI is not used at all in the font size calculation, and a default is used, as also pointed out by @amitdo - my test above just seems to verify that.

In my examples, the big font size is not at all reasonable since the page is in A4 format, and I would not expect the font size to be in pixels some of the time, and in actual font size at other times. Is it possible that the default dpi of 70 made you think the font size was in pixels? `x_fsize` should not be specified in pixels I think: http://kba.cloud/hocr-spec/1.2/#x_fsize

I think that the unit is defined, and the problem is that the unit cannot be interpreted without the DPI that was used, which is what I am suggesting to add.

**Comment by wollmers** [comment]
@MerlijnWajer 
> @wollmers - my understanding is that the estimated DPI is not used at all in the font size calculation, and a default is used, as also pointed out by @amitdo - my test above just seems to verify that.

Yes, I confirm. At 70 ppi the value of `x_fsize` is very similar to `x_size`. Sorry, my mistake was underestimating the difference between European metric points (0.376 mm) and US Printer Points (0,351367 mm) or DTP points (0.35278 mm). If I apply US PP to `x_size` I get exactly the value of  `x_fsize` (rounded integer).

**Comment by amitdo** [comment]
The PR will print something like "300 300". I think "300" is enough.

**Comment by MerlijnWajer** [comment]
The spec here seems to suggest it has to be two values: http://kba.cloud/hocr-spec/1.2/#scan_res - unless I am not reading it correctly?

**Comment by amitdo** [comment]
This feature was implemented in #3571.
